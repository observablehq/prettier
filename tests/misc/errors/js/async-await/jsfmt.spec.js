const { outdent } = require("outdent");

run_spec(
  {
    dirname: __dirname,
    snippets: [
      outdent`
        async function foo() {
          function bar(x = await 2) {}
        }
      `,
      "async (x = await 2) => {};",
      "function foo(promise) { await promise; }",
      outdent`
        function a() {
          return await 1
        }
      `,
      "() => { await x };",
      "C = class { static { function f() { await } } };",
      "C = class { static { function f(await) {} } };",
      "C = class { static { function f(x = await) {} } };",
      "C = class { static { function f({ [await]: x }) {} } };",
      "f = async (a) => await a! ** 6;",
      "f = (a) => +a! ** 6;",
      "async (a) => (await a!) ** 6;",
      "(-+5 ** 6);",
      // TODO: Test these two edge case
      // outdent`
      //   export class C {
      //     p = await (0);
      //   }
      // `,
      // "await (0);"
    ],
  },
  ["babel", "espree", "meriyah", "flow"]
);

run_spec(
  {
    dirname: __dirname,
    snippets: [
      outdent`
        async function foo() {
          function bar(x = await (2)) {}
        }
      `,
      outdent`
        function foo() {
          await
          foo;
        }
      `,
      "function foo(promise) { await (promise); }",
      outdent`
        function a() {
          return await (1)
        }
      `,
      "() => { await (x) };",
      outdent`
        function foo() {
          await
          (foo);
        }
      `,
      // `espree` and `meriyah` didn't throw
      "f = async () => await 5 ** 6;",
      // `espree` didn't throw
      "f = async () => await -5 ** 6;",
    ],
  },
  ["babel"]
);

run_spec(
  {
    dirname: __dirname,
    snippets: ["async (x = await (2)) => {};"],
  },
  ["babel", "espree", "meriyah"]
);

run_spec(
  {
    dirname: __dirname,
    snippets: [
      outdent`
        export class C {
          p = await 0;
        }
      `,
      "await 0;",
    ],
  },
  ["espree", "meriyah"]
);
