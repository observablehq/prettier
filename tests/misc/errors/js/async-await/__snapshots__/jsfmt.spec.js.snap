// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snippet: #0 [babel] format 1`] = `
"'await' is only allowed within async functions (2:20)
  1 | async function foo() {
> 2 |   function bar(x = await 2) {}
    |                    ^
  3 | }"
`;

exports[`snippet: #0 [babel] format 2`] = `
"Unexpected reserved word 'await' (2:20)
  1 | async function foo() {
> 2 |   function bar(x = await (2)) {}
    |                    ^
  3 | }"
`;

exports[`snippet: #0 [babel] format 3`] = `
"await is not allowed in async function parameters (1:12)
> 1 | async (x = await (2)) => {};
    |            ^"
`;

exports[`snippet: #0 [espree] format 1`] = `
"Cannot use keyword 'await' outside an async function (2:20)
  1 | async function foo() {
> 2 |   function bar(x = await 2) {}
    |                    ^
  3 | }"
`;

exports[`snippet: #0 [espree] format 2`] = `
"Cannot use keyword 'await' outside an async function (1:12)
> 1 | async (x = await (2)) => {};
    |            ^"
`;

exports[`snippet: #0 [espree] format 3`] = `
"Unexpected token = (2:5)
  1 | export class C {
> 2 |   p = await 0;
    |     ^
  3 | }"
`;

exports[`snippet: #0 [flow] format 1`] = `
"Unexpected number, expected the token \`,\` (2:26)
  1 | async function foo() {
> 2 |   function bar(x = await 2) {}
    |                          ^
  3 | }"
`;

exports[`snippet: #0 [meriyah] format 1`] = `
"[2:24]: 'Await' may not be used as an identifier in this context (2:24)
  1 | async function foo() {
> 2 |   function bar(x = await 2) {}
    |                        ^
  3 | }"
`;

exports[`snippet: #0 [meriyah] format 2`] = `
"[1:16]: 'Await' may not be used as an identifier in this context (1:16)
> 1 | async (x = await (2)) => {};
    |                ^"
`;

exports[`snippet: #0 [meriyah] format 3`] = `
"[2:11]: 'Await' may not be used as an identifier in this context (2:11)
  1 | export class C {
> 2 |   p = await 0;
    |           ^
  3 | }"
`;

exports[`snippet: #1 [babel] format 1`] = `
"await is not allowed in async function parameters (1:12)
> 1 | async (x = await 2) => {};
    |            ^"
`;

exports[`snippet: #1 [babel] format 2`] = `
"Unexpected reserved word 'await' (2:3)
  1 | function foo() {
> 2 |   await
    |   ^
  3 |   foo;
  4 | }"
`;

exports[`snippet: #1 [espree] format 1`] = `
"Cannot use keyword 'await' outside an async function (1:12)
> 1 | async (x = await 2) => {};
    |            ^"
`;

exports[`snippet: #1 [espree] format 2`] = `
"Cannot use keyword 'await' outside an async function (1:1)
> 1 | await 0;
    | ^"
`;

exports[`snippet: #1 [flow] format 1`] = `
"Unexpected number, expected the token \`,\` (1:18)
> 1 | async (x = await 2) => {};
    |                  ^"
`;

exports[`snippet: #1 [meriyah] format 1`] = `
"[1:16]: 'Await' may not be used as an identifier in this context (1:16)
> 1 | async (x = await 2) => {};
    |                ^"
`;

exports[`snippet: #1 [meriyah] format 2`] = `
"[1:5]: 'Await' may not be used as an identifier in this context (1:5)
> 1 | await 0;
    |     ^"
`;

exports[`snippet: #2 [babel] format 1`] = `
"'await' is only allowed within async functions (1:25)
> 1 | function foo(promise) { await promise; }
    |                         ^"
`;

exports[`snippet: #2 [babel] format 2`] = `
"Unexpected reserved word 'await' (1:25)
> 1 | function foo(promise) { await (promise); }
    |                         ^"
`;

exports[`snippet: #2 [espree] format 1`] = `
"Cannot use keyword 'await' outside an async function (1:25)
> 1 | function foo(promise) { await promise; }
    |                         ^"
`;

exports[`snippet: #2 [flow] format 1`] = `
"Unexpected identifier, expected the end of an expression statement (\`;\`) (1:31)
> 1 | function foo(promise) { await promise; }
    |                               ^^^^^^^"
`;

exports[`snippet: #2 [meriyah] format 1`] = `
"[1:29]: 'Await' may not be used as an identifier in this context (1:29)
> 1 | function foo(promise) { await promise; }
    |                             ^"
`;

exports[`snippet: #3 [babel] format 1`] = `
"'await' is only allowed within async functions (2:10)
  1 | function a() {
> 2 |   return await 1
    |          ^
  3 | }"
`;

exports[`snippet: #3 [babel] format 2`] = `
"Unexpected reserved word 'await' (2:10)
  1 | function a() {
> 2 |   return await (1)
    |          ^
  3 | }"
`;

exports[`snippet: #3 [espree] format 1`] = `
"Cannot use keyword 'await' outside an async function (2:10)
  1 | function a() {
> 2 |   return await 1
    |          ^
  3 | }"
`;

exports[`snippet: #3 [flow] format 1`] = `
"Unexpected number, expected the token \`;\` (2:16)
  1 | function a() {
> 2 |   return await 1
    |                ^
  3 | }"
`;

exports[`snippet: #3 [meriyah] format 1`] = `
"[2:14]: 'Await' may not be used as an identifier in this context (2:14)
  1 | function a() {
> 2 |   return await 1
    |              ^
  3 | }"
`;

exports[`snippet: #4 [babel] format 1`] = `
"'await' is only allowed within async functions (1:9)
> 1 | () => { await x };
    |         ^"
`;

exports[`snippet: #4 [babel] format 2`] = `
"Unexpected reserved word 'await' (1:9)
> 1 | () => { await (x) };
    |         ^"
`;

exports[`snippet: #4 [espree] format 1`] = `
"Cannot use keyword 'await' outside an async function (1:9)
> 1 | () => { await x };
    |         ^"
`;

exports[`snippet: #4 [flow] format 1`] = `
"Unexpected identifier, expected the end of an expression statement (\`;\`) (1:15)
> 1 | () => { await x };
    |               ^"
`;

exports[`snippet: #4 [meriyah] format 1`] = `
"[1:13]: 'Await' may not be used as an identifier in this context (1:13)
> 1 | () => { await x };
    |             ^"
`;

exports[`snippet: #5 [babel] format 1`] = `
"Unexpected reserved word 'await' (1:37)
> 1 | C = class { static { function f() { await } } };
    |                                     ^"
`;

exports[`snippet: #5 [babel] format 2`] = `
"Unexpected reserved word 'await' (2:3)
  1 | function foo() {
> 2 |   await
    |   ^
  3 |   (foo);
  4 | }"
`;

exports[`snippet: #5 [espree] format 1`] = `
"Unexpected token { (1:20)
> 1 | C = class { static { function f() { await } } };
    |                    ^"
`;

exports[`snippet: #5 [flow] format 1`] = `
"Unexpected token \`{\`, expected an identifier (1:20)
> 1 | C = class { static { function f() { await } } };
    |                    ^"
`;

exports[`snippet: #5 [meriyah] format 1`] = `
"[1:20]: Unexpected token: '{' (1:20)
> 1 | C = class { static { function f() { await } } };
    |                    ^"
`;

exports[`snippet: #6 [babel] format 1`] = `
"Unexpected reserved word 'await' (1:33)
> 1 | C = class { static { function f(await) {} } };
    |                                 ^"
`;

exports[`snippet: #6 [babel] format 2`] = `
"Illegal expression. Wrap left hand side or entire exponentiation in parentheses. (1:23)
> 1 | f = async () => await 5 ** 6;
    |                       ^"
`;

exports[`snippet: #6 [espree] format 1`] = `
"Unexpected token { (1:20)
> 1 | C = class { static { function f(await) {} } };
    |                    ^"
`;

exports[`snippet: #6 [flow] format 1`] = `
"Unexpected token \`{\`, expected an identifier (1:20)
> 1 | C = class { static { function f(await) {} } };
    |                    ^"
`;

exports[`snippet: #6 [meriyah] format 1`] = `
"[1:20]: Unexpected token: '{' (1:20)
> 1 | C = class { static { function f(await) {} } };
    |                    ^"
`;

exports[`snippet: #7 [babel] format 1`] = `
"Unexpected reserved word 'await' (1:37)
> 1 | C = class { static { function f(x = await) {} } };
    |                                     ^"
`;

exports[`snippet: #7 [babel] format 2`] = `
"Illegal expression. Wrap left hand side or entire exponentiation in parentheses. (1:23)
> 1 | f = async () => await -5 ** 6;
    |                       ^"
`;

exports[`snippet: #7 [espree] format 1`] = `
"Unexpected token { (1:20)
> 1 | C = class { static { function f(x = await) {} } };
    |                    ^"
`;

exports[`snippet: #7 [flow] format 1`] = `
"Unexpected token \`{\`, expected an identifier (1:20)
> 1 | C = class { static { function f(x = await) {} } };
    |                    ^"
`;

exports[`snippet: #7 [meriyah] format 1`] = `
"[1:20]: Unexpected token: '{' (1:20)
> 1 | C = class { static { function f(x = await) {} } };
    |                    ^"
`;

exports[`snippet: #8 [babel] format 1`] = `
"Unexpected reserved word 'await' (1:36)
> 1 | C = class { static { function f({ [await]: x }) {} } };
    |                                    ^"
`;

exports[`snippet: #8 [espree] format 1`] = `
"Unexpected token { (1:20)
> 1 | C = class { static { function f({ [await]: x }) {} } };
    |                    ^"
`;

exports[`snippet: #8 [flow] format 1`] = `
"Unexpected token \`{\`, expected an identifier (1:20)
> 1 | C = class { static { function f({ [await]: x }) {} } };
    |                    ^"
`;

exports[`snippet: #8 [meriyah] format 1`] = `
"[1:20]: Unexpected token: '{' (1:20)
> 1 | C = class { static { function f({ [await]: x }) {} } };
    |                    ^"
`;

exports[`snippet: #9 [babel] format 1`] = `
"Unexpected token (1:27)
> 1 | f = async (a) => await a! ** 6;
    |                           ^"
`;

exports[`snippet: #9 [espree] format 1`] = `
"Unexpected token ! (1:25)
> 1 | f = async (a) => await a! ** 6;
    |                         ^"
`;

exports[`snippet: #9 [flow] format 1`] = `
"Unexpected token \`!\`, expected the end of an expression statement (\`;\`) (1:25)
> 1 | f = async (a) => await a! ** 6;
    |                         ^"
`;

exports[`snippet: #9 [meriyah] format 1`] = `
"[1:25]: Unexpected token: '!' (1:25)
> 1 | f = async (a) => await a! ** 6;
    |                         ^"
`;

exports[`snippet: #10 [babel] format 1`] = `
"Unexpected token (1:16)
> 1 | f = (a) => +a! ** 6;
    |                ^"
`;

exports[`snippet: #10 [espree] format 1`] = `
"Unexpected token ! (1:14)
> 1 | f = (a) => +a! ** 6;
    |              ^"
`;

exports[`snippet: #10 [flow] format 1`] = `
"Unexpected token \`!\`, expected the end of an expression statement (\`;\`) (1:14)
> 1 | f = (a) => +a! ** 6;
    |              ^"
`;

exports[`snippet: #10 [meriyah] format 1`] = `
"[1:14]: Unexpected token: '!' (1:14)
> 1 | f = (a) => +a! ** 6;
    |              ^"
`;

exports[`snippet: #11 [babel] format 1`] = `
"Unexpected token, expected \\",\\" (1:22)
> 1 | async (a) => (await a!) ** 6;
    |                      ^"
`;

exports[`snippet: #11 [espree] format 1`] = `
"Unexpected token ! (1:22)
> 1 | async (a) => (await a!) ** 6;
    |                      ^"
`;

exports[`snippet: #11 [flow] format 1`] = `
"Unexpected token \`!\`, expected the token \`)\` (1:22)
> 1 | async (a) => (await a!) ** 6;
    |                      ^"
`;

exports[`snippet: #11 [meriyah] format 1`] = `
"[1:22]: Expected ')' (1:22)
> 1 | async (a) => (await a!) ** 6;
    |                      ^"
`;

exports[`snippet: #12 [babel] format 1`] = `
"Illegal expression. Wrap left hand side or entire exponentiation in parentheses. (1:3)
> 1 | (-+5 ** 6);
    |   ^"
`;

exports[`snippet: #12 [espree] format 1`] = `
"Unexpected token ** (1:6)
> 1 | (-+5 ** 6);
    |      ^"
`;

exports[`snippet: #12 [flow] format 1`] = `
"Invalid left-hand side in exponentiation expression (1:2)
> 1 | (-+5 ** 6);
    |  ^^^"
`;

exports[`snippet: #12 [meriyah] format 1`] = `
"[1:7]: Unary expressions as the left operand of an exponentiation expression must be disambiguated with parentheses (1:7)
> 1 | (-+5 ** 6);
    |       ^"
`;
